src/file.rs

```diff
@@ -205,71 +238,92 @@ impl AsepriteFile {
-        let blend_fn = blend_mode_to_blend_fn(layer.blend_mode());
-        match &cel.data {
-            CelData::Linked(frame) => {
-                for cel in self.framedata.cel(*frame, cel.layer_index) {
-                    match &cel.data {
-                        CelData::Linked(_) => {
-                            panic!("Cel links to empty cel. Should have been caught by validate()");
-                        }
-                        CelData::RawRgba {
-                            width,
-                            height,
-                            data,
-                        } => {
-                            copy_cel_to_image(
-                                image,
-                                cel.x as i32,
-                                cel.y as i32,
-                                *width as i32,
-                                *height as i32,
-                                cel.opacity,
-                                &data.0,
-                                &blend_fn,
-                            );
-                        }
-                        CelData::RawIndexed { .. } => {
-                            panic!("Indexed data cel. Should have been caught by validate()");
-                        }
+        let RawCel { data, content } = cel;
+        let layer = self.layer(data.layer_index as u32);
+        let blend_mode = layer.blend_mode();
+        match &content {
+            CelContent::Raw(image_content) => {
+                let ImageContent { size, pixels } = image_content;
+                match pixels {
+                    pixel::Pixels::Rgba(pixels) => {
+                        write_raw_cel_to_image(image, data, size, pixels, &blend_mode);
+                    }
+                    pixel::Pixels::Grayscale(_) => {
+                        panic!("Grayscale cel. Should have been caught by validate()");
+                    }
+                    pixel::Pixels::Indexed(_) => {
+                        panic!("Indexed data cel. Should have been caught by validate()");
                     }
                 }
             }
-            CelData::RawRgba {
-                width,
-                height,
-                data,
-            } => {
-                copy_cel_to_image(
-                    image,
-                    cel.x as i32,
-                    cel.y as i32,
-                    *width as i32,
-                    *height as i32,
-                    cel.opacity,
-                    &data.0,
-                    &blend_fn,
-                );
+            CelContent::Tilemap(tilemap_data) => {
+                let layer_type = layer.layer_type();
+                let tileset_id = layer_type.expect_tilemap();
+                let tileset = &self.tilesets()[*tileset_id];
+                let tileset_pixels = tileset
+                    .pixels
+                    .as_ref()
+                    .expect("Expected Tileset data to contain pixels");
```

Comment:
For all these `expect` calls: They either need to be verified by the validator. Or we should print a `warn!` and just do nothing instead. But we shouldn't crash/panic.